File: SortVision/public/code/radix/julia/radixSort.jl
# Radix Sort (LSD, base 10, non-negative integers)
# Time: O(d * (n + k)), Space: O(n + k)
function radix_sort!(arr::Vector{Int})
    if length(arr) <= 1
        return arr
    end

    max_val = maximum(arr)
    exp = 1
    while div(max_val, exp) > 0
        counting_sort_by_exp!(arr, exp)
        exp *= 10
    end
    return arr
end

function counting_sort_by_exp!(arr::Vector{Int}, exp::Int)
    output = similar(arr)
    count = fill(0, 10)

    for x in arr
        count[(div(x, exp) % 10) + 1] += 1
    end
    for i in 2:10
        count[i] += count[i - 1]
    end

    for i in length(arr):-1:1
        digit = (div(arr[i], exp) % 10) + 1
        output[count[digit]] = arr[i]
        count[digit] -= 1
    end

    arr .= output
end

arr = [170, 45, 75, 90, 802, 24, 2, 66]
println(radix_sort!(arr))

File: SortVision/public/code/radix/rust/radixSort.rs
// Radix Sort (LSD, base 10, non-negative integers)
// Time: O(d * (n + k)), Space: O(n + k)
pub fn radix_sort(arr: &mut [u32]) {
    if arr.len() <= 1 {
        return;
    }

    let mut max_val = arr[0];
    for &x in arr.iter() {
        if x > max_val {
            max_val = x;
        }
    }

    let mut exp = 1u32;
    while max_val / exp > 0 {
        counting_sort_by_exp(arr, exp);
        exp *= 10;
    }
}

fn counting_sort_by_exp(arr: &mut [u32], exp: u32) {
    let mut output = vec![0u32; arr.len()];
    let mut count = [0usize; 10];

    for &x in arr.iter() {
        count[((x / exp) % 10) as usize] += 1;
    }
    for i in 1..10 {
        count[i] += count[i - 1];
    }

    for i in (0..arr.len()).rev() {
        let digit = ((arr[i] / exp) % 10) as usize;
        output[count[digit] - 1] = arr[i];
        count[digit] -= 1;
    }

    arr.copy_from_slice(&output);
}

fn main() {
    let mut arr = [170, 45, 75, 90, 802, 24, 2, 66];
    radix_sort(&mut arr);
    println!("{:?}", arr);
}
